nab/detectors/htmcore/htmcore_detector.py

```diff
@@ -332,6 +334,12 @@ def modelRun(self, ts, val):
           pass
 
       if PANDA_VIS_ENABLED:
+          # activateDendrites calculates active segments
+          self.tm.activateDendrites(learn=True)
+          # predictive cells are calculated directly from active segments
+          predictiveCells = self.tm.getPredictiveCells()
```

Comment:
nit: does this not work for all cases (I mean including the 1st)?. `activate dendrites & then cells`

Reply:
Yeah, it works, but things get complicated, when you consider inputs. And time consistency.

--script start
Inputs are in time T.
Active columns are in time T.
Winners (active cells) are in time T.
Prediction cells must be in T+1.
--pandaBlockExecution
--endscript

And for vizualization pruposes in need this order if i want to prevent buffering previous values etc..
Also while using external distal inputs for TM this order is OK.

Reply:
> And for vizualization pruposes in need this order if i want to prevent buffering previous values etc..
Also while using external distal inputs for TM this order is OK. 

see above - predictive inputs shouldn't be a problem :heavy_check_mark: , but I guess the order would be `dendrites, then activate cells` (and then you get predictive at T ~ "active T+1"), so you need to do the buffering (can the visualizer do that internally?, as this is true for all HTMs.

Reply:
If i go with the path of buffering, there are several issues.

Vis has besides "do one step" also "GOTO iteration xxx" and if i will buffer, i need to stop one step before goal iteration and exectute that, send data and then move on to the destination iteration.

There is also the feature for requesting the proximal/distal synapses on user-mouse-click-demand for specific cell/column. If i am not wrong, activateCells grows/penalize synapses, so the state of synapses would be T+1 and again i would need to buffer that... but i can't know what user will want :/

Because these are bigdata - visualizer only requests data like synapses for specific cell/column that user mouse clicks.
I can't send all data, that would be computationally intensive...

Maybe in the future i would like to improve this to log historical data and keep track of some focused columns/cells, that would be awesome.. but for now i need to stay on the ground

Reply:
> Vis has besides "do one step" also "GOTO iteration xxx" and if i will buffer, i need to stop one step before goal iteration and exectute tha 

the implementation gets complicated, esp to support the "jumps". But..are the data/is the approach correct? Could you least empirically verify (with fixed seed) that the result state from `compute()` vs "suggested approach here" is identical?

Reply:
> If i am not wrong, activateCells grows/penalize synapses, so the state of synapses would be T+1 and again i would need to buffer that... but i can't know what user will want :/ 

you're right with the penalization. `activateDendrites-activateCells` action-pair gives you "active cells for T", and "predictive cells for T" (which approximates state at T+1) 

> Maybe in the future i would like to improve this to log historical data and keep track of some focused columns/cells, that would be awesome.. but for now i need to stay on the ground 

that would be cool! some video,backtrack feature for select cells/patterns.